### PR TITLE
MSC2790 Modal Widgets

### PR DESCRIPTION
**Pull request by t3chguy**
https://github.com/matrix-org/matrix-doc/blob/travis/msc/modal-widgets/proposals/2790-modal-widgets.md

### CODE REVIEW

**Review by turt2live — CHANGES_REQUESTED** [review]
generally looks good from this side though - thanks!

File: src/ClientWidgetApi.ts [hunk 1]
```diff
@@ -193,4 +199,20 @@ export class ClientWidgetApi extends AlmostEventEmitter {
             visible: isVisible,
         });
     }
+
+    public sendWidgetConfig(data: IWidgetData): Promise<void> {
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/ClientWidgetApi.ts:211]: That's assuming that widget_config is only for Modal Widgets, if we choose to keep it generic then it should be any possible WidgetData as it sends `definition.data`

[the remark this answers is not in the document]
Comment (turt2live) [reply src/ClientWidgetApi.ts:211]: it's supposed to send `definition`.

[the remark this answers is not in the document]
Comment (t3chguy) [reply src/ClientWidgetApi.ts:211]: TS doesn't like it

image

[the remark this answers is not in the document]
Comment (t3chguy) [reply src/ClientWidgetApi.ts:211]: I still think it should be IWidgetData

> The widget_config action has a data copied from the original open_modal action sent by the source widget. The client should strip any keys off the modal widget definition which are not covered by this MSC - the source widget is expected to only use the data property of the widget definition, not the root of the widget definition, to pass data through.

IWidget has creatorUserId and is missing buttons from the type. `IModalWidgetOpenRequestData` seems more appropriate

**Review by turt2live — APPROVED** [review]

**Pull request merged**